airbyte-ci/connectors/auto_merge/src/auto_merge/main.py

```diff
@@ -0,0 +1,178 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+from __future__ import annotations
+
+import logging
+import os
+import time
+from collections.abc import Iterable, Iterator
+from contextlib import contextmanager
+from typing import TYPE_CHECKING
+
+from github import Auth, Github
+
+from .consts import AIRBYTE_REPO, AUTO_MERGE_LABEL, BASE_BRANCH, CONNECTOR_PATH_PREFIXES
+from .env import GITHUB_TOKEN, PRODUCTION
+
+if TYPE_CHECKING:
+    from github.Commit import Commit as GithubCommit
+    from github.File import File as GithubFile
+    from github.PullRequest import PullRequest
+    from github.Repository import Repository as GithubRepo
+
+logging.basicConfig()
+logging.getLogger().setLevel(logging.INFO)
+
+
+@contextmanager
+def github_client() -> Iterator[Github]:
+    client = None
+    try:
+        client = Github(auth=Auth.Token(GITHUB_TOKEN), seconds_between_requests=0)
+        yield client
+    finally:
+        if client:
+            client.close()
+
+
+def check_if_modifies_connectors_only(modified_files: Iterable[GithubFile]) -> bool:
+    """Check if all modified files are in CONNECTOR_PATH_PREFIXES
+
+    Args:
+        modified_files (Iterable[GithubFile]): All the modified files on a PR.
+
+    Returns:
+        bool: True if all modified files are in CONNECTOR_PATH_PREFIXES, False otherwise
+    """
+    for file in modified_files:
+        if not any(file.filename.startswith(prefix) for prefix in CONNECTOR_PATH_PREFIXES):
+            return False
+    return True
+
+
+def check_if_head_commit_passes_all_required_checks(head_commit: GithubCommit, required_checks: set[str]) -> bool:
+    """Required checks can be a mix of status contexts and check runs. A head commit is considered to pass all required checks if it has successful statuses and check runs for all required checks.
+
+    Args:
+        head_commit (GithubCommit): The head commit of the PR
+        required_checks (set[str]): The set of required passing checks
+
+    Returns:
+        bool: True if the head commit passes all required checks, False otherwise
+    """
+    successful_status_contexts = [commit_status.context for commit_status in head_commit.get_statuses() if commit_status.state == "success"]
+    successful_check_runs = [check_run.name for check_run in head_commit.get_check_runs() if check_run.conclusion == "success"]
+    successful_contexts = set(successful_status_contexts + successful_check_runs)
+    return required_checks.issubset(successful_contexts)
+
+
+def check_if_pr_is_auto_mergeable(head_commit: GithubCommit, pr: PullRequest, required_checks: set[str]) -> bool:
```

Comment:
💡 Ok so here is the most important logic in the system. The business logic.

Its important that its
1. Readable
2. Easily extendable
3. Contained to as few locations (files) as possible

I think youve achieved that so this is suggestion is **very** optional:

What if we used a validator pattern


```python
def _validate_has_auto_merge_label((head_commit: GithubCommit, pr: PullRequest, required_checks: set[str]) -> bool, Optional[str]:
    has_auto_merge_label = any(label.name == AUTO_MERGE_LABEL for label in pr.labels)
    if not has_auto_merge_label:
        return False, f"PR {pr.number} does not have the {AUTO_MERGE_LABEL} label"

    return True, None
    
def check_if_pr_is_auto_mergeable(head_commit: GithubCommit, pr: PullRequest, required_checks: set[str]) -> bool:
    validators = [
        _validate_has_auto_merge_label,
        # ...
    ]

    for pr_validator in validators:
        is_valid, error = pr_validator(head_commit, pr, required_checks)        
        if not is_valid
            logging.info(error)
            return False
```

_(this is an example and could be improved (Named `ValidationArgument|Result` type, standard error messaging, etc) but the ideas there.)_

Why this way?

1. It lets us conditionally add/remove validations for a connector more easily. You can imagine that critical connectors or third party connectors might want to add or skip certain validations
2. Validation logic is easier to test
3. imo easier to skim and understand what validations run when/under what conditions

Anyway again very optional